src/migrations/20221010162913-add-self-trade-type.ts

```diff
@@ -0,0 +1,186 @@
+
+  const nonUserWalletIds = []
+  const walletsCursor = walletsCollection.find({
+    _accountId: { $in: accountObjectIds },
+  })
+  for await (const wallet of walletsCursor) {
+    nonUserWalletIds.push(wallet.id)
+  }
+  return nonUserWalletIds
+}
+
+module.exports = {
+  async up(db) {
+    const usersCollection = db.collection("users")
+    const walletsCollection = db.collection("wallets")
+    const txnsCollection = db.collection("medici_transactions")
+
+    const nonUserWalletIds = await getNonUserWalletIds({
+      usersCollection,
+      walletsCollection,
+    })
+
+    const accountIdsByWalletId = {}
+    const getSelfTradeType = async (txnAgg) => {
+      // BUILD 'txns' FOR JOURNALID
+      // =====
+      const txns = []
+      const txnsCursor = txnsCollection.find({ _journal: txnAgg._id })
+      for await (const txn of txnsCursor) {
+        // Filter internal wallet IDs
+        if (nonUserWalletIds.includes(toWalletId(txn.accounts))) continue
+
+        // Filter non-user walletIds
+        const walletId = toWalletId(txn.accounts)
+        if (!walletId) continue
+
+        // Fetch accountId locally
+        let accountId = accountIdsByWalletId[walletId]
+        if (accountId) {
+          txns.push({ walletId, currency: txn.currency, accountId })
+          continue
+        }
+
+        // Fetch accountId from db if not cached
+        const wallet = await walletsCollection.findOne({ id: walletId })
+        if (!wallet) continue
+        accountId = wallet._accountId.toString()
+        if (!accountId) continue
+
+        accountIdsByWalletId[walletId] = accountId
+        txns.push({ walletId, currency: txn.currency, accountId })
+      }
+
+      // CHECK 'txns' TO DETERMINE IF SELF-TRADE
+      // =====
+
+      // Check that we have txns with valid walletIds
+      if (txns && txns.length === 0) return false
+
+      // Check if we have exactly 2 txns
+      if (txns.length !== 2) return false
+
+      // Check if txns have different currencies
+      if (txns[0].currency === txns[1].currency) return false
+
+      // Compare accountIds to see if self-trade
+      const accountIdsSet = new Set(txns.map((txn) => txn.accountId))
+      if (accountIdsSet.size > 1) return false
+
+      // RETURN SELF-TRADE TYPE
+      // =====
+      switch (txnAgg.type) {
+        case LedgerTransactionType.IntraLedger:
+          return LedgerTransactionType.WalletIdTradeIntraAccount
+        case LedgerTransactionType.LnIntraLedger:
+          return LedgerTransactionType.LnTradeIntraAccount
+        case LedgerTransactionType.OnchainIntraLedger:
+          return LedgerTransactionType.OnChainTradeIntraAccount
+        default:
+          return false
+      }
+    }
+
+    // FETCH INTRALEDGER TXNS AND UPDATE TYPE FOR SELF-TRADES
+    // =====
+    const journalCursor = txnsCollection.aggregate([
+      {
+        $match: { type: { $regex: "on_us" }, currency: "USD" },
```

Comment:
is this always `on_us`? cant be `ln_on_us` or `onchain_on_us`?

Reply:
Regex pattern catches all of the above